### PR TITLE
Add ExtraCode declarationFile and implementationFile directives

### PR DESCRIPTION
**Pull request by m-fila**
BEGINRELEASENOTES
- Added  ExtraCode declarationFile and implementationFile directives

ENDRELEASENOTES

Added 'ExtraCode'  directives 'declarationFile' and 'implementationFile' used to include/embed files in code generated with jinja. This feature was planned but not implemented so far. Can be used to fix https://github.com/key4hep/EDM4hep/issues/296

The paths to 'declarationFile' and 'implementationFile 'can be either absolute or relative to datamodel yaml. Should this be relative to working directory?

The files declared in 'declarationFile' and 'implementationFile' are not added as configuration dependency in cmake, so the changes to these files will not automatically trigger re-generating a data model. I don't know if there is an easy way to fix this

TODO:
- [x] test mutable declaration and mutable implementations in datatype
- [x] add component 'declarationFile'
- [x] test declarationFile in component
- [x] fix tests that copy the yml file

### CODE REVIEW

**Review by tmadlener — COMMENTED** [review]
Thanks for taking this on. IIUC the injection happens in the jinja stage, right? This also means that the implementation files (or at least their contents) are not stored in the output files, right? (At least that is my explanation for having to carry them to the tests directory for the roundtrip tests).

Would it be possible to have the "parsing" of these implementation file at the datamodel reading stage? I.e. whenever we encounter such an implementation file, we just inject the contents directly into the existing `ExtraCode` bit? In that way we they should directly end up in the datamodel definition that is stored in the file (I think / hope). Additionally, I think we would not have to touch the generation step at all(?).

**Comment by m-fila** [comment]
Yes, during jinja step a python function is used to read a file and return its content to jinja, which just put it as is into generated file. The files itself are not copied anywhere.
In the metada datamodel definitions the 'declarationFile' and 'implementationFile' are put without any modifications, which was causing problems when the paths are relative, hence I copied the files to build directory

I think appending the contents of 'declatationFile' to 'declaration' during parsing should work better and give more transparent metadata. Thanks :smiley:

**Review by tmadlener — COMMENTED** [review]
Very nice. Did you check whether this works in the end for EDM4hep and whether it really resolves the issue you set out to fix?

_____ 
> The paths to 'declarationFile' and 'implementationFile 'can be either absolute or relative to datamodel yaml. Should this be relative to working directory?

No I think absolute, or relative to the yaml file is fine. But we should document it.

> The files declared in 'declarationFile' and 'implementationFile' are not added as configuration dependency in cmake, so the changes to these files will not automatically trigger re-generating a data model. I don't know if there is an easy way to fix this

One possibility (that would make it easy for us / podio again) would be to document this and tell people that they have to add this to their cmake configuration via, similar to what we do in the macro:
https://github.com/AIDASoft/podio/blob/a18229b9b1f1db6530dec0bd7c6e48f26a0a5f9f/cmake/podioMacros.cmake#L201-L211

The only thing we could potentially offer is to make the macro take an additional list of the files that are used in the datamodel and add them there. But at that point the users do not gain too much, I think.

File: python/podio_gen/podio_config_reader.py [hunk 1]
```diff
@@ -442,8 +449,20 @@ def _handle_extracode(definition):
...
+    def _expand_extracode(definitions, parent_path, directive):
+        """Expand extra code directives by including files from 'File' directives
+        Relative paths to files are extended with parent_path. Mutates definitions"""
+        if directive + "File" in definitions.keys():
+            filename = definitions.pop(directive + "File")
+            if not os.path.isabs(filename):
+                filename = os.path.join(parent_path, filename)
+            with open(filename, "r", encoding="utf-8") as stream:
+                contents = stream.read()
+            definitions[directive] = definitions.get(directive, "") + "\n" + contents
```
[the remark this answers is not in the document]
Comment (m-fila) [reply python/podio_gen/podio_config_reader.py:462]: Yes, formatting is preserved. In a dumped model it becomes:
```
      declaration: " int negate() { x = -x; return x; }\n \nint reset() {\n  x = 0;\n\
        \  return x;\n}\nint get() {\n  return 42;\n}"
```

File: python/podio_gen/podio_config_reader.py [hunk 2]
```diff
@@ -171,9 +172,15 @@ class ClassDefinitionValidator:
...
+    valid_extra_code_keys = (
+        "declaration",
+        "implementation",
+        "includes",
+        "declarationFile",
+        "implementationFile",
+    )
     # documented but not yet implemented
-    not_yet_implemented_extra_code = ("declarationFile", "implementationFile")
+    not_yet_implemented_extra_code = ()
```
[the remark this answers is not in the document]
Comment (m-fila) [reply python/podio_gen/podio_config_reader.py]: removed

File: doc/datamodel_syntax.md [hunk 3]
```diff
@@ -119,17 +119,17 @@ The `includes` will be add to the header files of the generated classes.
...
       declaration: <string>
       implementation : <string>
-      declarationFile: <string> (to be implemented!)
-      implementationFile: <string> (to be implemented!)
+      declarationFile: <string>
+      implementationFile: <string>
 ```
 
-The code being provided has to use the macro `{name}` in place of the concrete name of the class.
+The code being provided has to use the macro `{name}` in place of the concrete name of the class. The paths to the files given in `declarationFile` and `implementationFile` should be either absolute or relative to the datamodel yaml file. The cmake users are encouraged to specify these file in `DEPENDS` option of `PODIO_GENERATE_DATAMODEL` to add the files as configuration dependency of the datamodel and cause the datamodel re-generation when they change.
```
Comment (tmadlener) [inline doc/datamodel_syntax.md:132]:
```suggestion
The code being provided has to use the macro `{name}` in place of the concrete name of the class. The paths to the files given in `declarationFile` and `implementationFile` should be either absolute or relative to the datamodel yaml file. The cmake users are encouraged to specify these file via the `DEPENDS` option of `PODIO_GENERATE_DATAMODEL` to add the files as configuration dependency of the datamodel and cause the datamodel re-generation when they change.
```

**Review by tmadlener — APPROVED** [review]

**Comment by hegner** [comment]
If builds run through, I will merge it

**Pull request merged**